### ISSUE TITLE
Wrong delete call in coda-grib.c

### ISSUE THREAD

**Issue by schwehr** [stcorp/coda#44]
Using

```c++
extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
  FuzzerTemporaryFile temp_file(data, size);

  coda_init();
  const char *product_class = NULL;
  const char *product_type = NULL;
  coda_format format;
  int version;
  coda_recognize_file(temp_file.filename(), NULL, &format, &product_class,
                      &product_type, &version);
  coda_done();
  return 0;
}

```

I got this crash:

```
0x0000555556534acf in __assert_fail (assertion=<optimized out>, file=<optimized out>, line=62, 
    function=0x555555928ae0 <__PRETTY_FUNCTION__.coda_grib_type_delete> "void coda_grib_type_delete(coda_dynamic_type *)") at base/logging.cc:106
#8  0x0000555555cf2bfc in coda_grib_type_delete (type=0x604000002050) at third_party/stcorp_coda/libcoda/coda-grib-type.c:62
#9  0x0000555555ce8ef5 in read_grib1_message (product=0x607000000560, message=<optimized out>, file_offset=<optimized out>) at third_party/stcorp_coda/libcoda/coda-grib.c:1727
#10 0x0000555555cdc85e in coda_grib_reopen (product=<optimized out>) at third_party/stcorp_coda/libcoda/coda-grib.c:3103
#11 0x0000555555d10a77 in reopen_with_backend (product_file=0x10418, format=66584) at third_party/stcorp_coda/libcoda/coda-product.c:408
#12 0x0000555555d0e61e in open_file (filename=<optimized out>, product_file=<optimized out>, force_binary=<optimized out>) at third_party/stcorp_coda/libcoda/coda-product.c:550
#13 0x0000555555d0e09d in coda_recognize_file (filename=<optimized out>, file_size=<optimized out>, file_format=<optimized out>, product_class=<optimized out>, 
    product_type=<optimized out>, version=<optimized out>) at third_party/stcorp_coda/libcoda/coda-product.c:594
#14 0x0000555555c68ced in LLVMFuzzerTestOneInput (data=<optimized out>, size=<optimized out>) at third_party/stcorp_coda/fuzz/coda_recognize_file_fuzzer.cc:16
```

The assert is here:

```c
void coda_grib_type_delete(coda_dynamic_type *type)
{
    assert(type != NULL);
    assert(type->backend == coda_backend_grib);  // <--- this assert was hit
```

Fixed by switching `coda_grib_type_delete` to `coda_mem_type_delete` lets the fuzzer poc run without crashing.

```c
            /* data representation type is Latitude/Longitude Grid */
            gds = coda_mem_record_new((coda_type_record *)grib_type[grib1_grid], NULL);

            NV = buffer[3];
            gtype = grib_type[grib1_numberOfVerticalCoordinateValues];
            type = (coda_dynamic_type *)coda_mem_uint8_new((coda_type_number *)gtype, NULL, cproduct, NV);
            coda_mem_record_add_field(gds, "numberOfVerticalCoordinateValues", type, 0);

            PVL = buffer[4];

            gtype = grib_type[grib1_dataRepresentationType];
            type = (coda_dynamic_type *)coda_mem_uint8_new((coda_type_number *)gtype, NULL, cproduct, buffer[5]);
            coda_mem_record_add_field(gds, "dataRepresentationType", type, 0);

            if (read_bytes(product->raw_product, file_offset, 26, buffer) < 0)
            {
                coda_mem_type_delete((coda_dynamic_type *)gds);
                // coda_grib_type_delete((coda_dynamic_type *)gds);  <----- wrong type
                return -1;
            }
```

**Comment by schwehr** [comment]
Fuzzer proof of concept (poc):

[crash-65723607e49c0ccb8ac6e7790f748cf607880bc9.zip](https://github.com/stcorp/coda/files/4463993/crash-65723607e49c0ccb8ac6e7790f748cf607880bc9.zip)

**Comment by schwehr** [comment]
The bug is in read_grib1_message()

**Comment by svniemeijer** [comment]
This is indeed a bug. Thanks for finding it.
Luckily it is in a cleanup handler (when you have e.g. corrupt grib files), so users should generally not be impacted.
The correct fix is to use `coda_dynamic_type_delete` though.
Fixed in 836d1493d7b70451ae9239b78e07d767da1d3adc
